crates/holochain_state/src/source_chain.rs

```diff
@@ -685,8 +681,69 @@ impl SourceChain {
                 }
                 SourceChainResult::Ok(())
             })
-            .await?;
-        Ok(())
+            .await
+        {
+            Err(SourceChainError::HeadMoved(
+                old_head,
+                Some((new_persisted_head, new_head_seq, new_timestamp)),
+            )) => {
+                let is_relaxed =
+                    self.scratch
+                        .apply_and_then::<bool, SyncScratchError, _>(|scratch| {
+                            Ok(scratch.chain_top_ordering() == ChainTopOrdering::Relaxed)
+                        })?;
+                if is_relaxed {
+                    let keystore = self.vault.keystore();
+                    async fn rebase_headers_on(
```

Comment:
All these internal functions make the flush function really hard to read. Could we please move them out of the flush function?